File: SkyEditor.Core/IO/SavableFileSaver.vb
Namespace IO
    ''' <summary>
    ''' A class that can save models that implement ISavable or ISavableAs
    ''' </summary>
    Public Class SavableFileSaver
        Implements IFileSaver

        Public Function GetDefaultExtension(model As Object) As String Implements IFileSaver.GetDefaultExtension
            Return DirectCast(model, ISavableAs).GetDefaultExtension()
        End Function

        Public Function Save(model As Object, provider As IOProvider) As Task Implements IFileSaver.Save
            DirectCast(model, ISavable).Save(provider)
            Return Task.FromResult(0)
        End Function

        Public Function Save(model As Object, filename As String, provider As IOProvider) As Task Implements IFileSaver.Save
            DirectCast(model, ISavableAs).Save(filename, provider)
            Return Task.FromResult(0)
        End Function

        Public Function SupportsSave(model As Object) As Boolean Implements IFileSaver.SupportsSave
            Return TypeOf model Is ISavable
        End Function

        Public Function SupportsSaveAs(model As Object) As Boolean Implements IFileSaver.SupportsSaveAs
            Return TypeOf model Is ISavableAs
        End Function
    End Class
End Namespace

File: SkyEditor.Core/IO/PluginInfrastructure/SavableFileSaver.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace SkyEditor.Core.IO.PluginInfrastructure
{
    public class SavableFileSaver : IFileSaver
    {
        public string GetDefaultExtension(object model)
        {
            return (model as ISavableAs)?.GetDefaultExtension();
        }

        public IEnumerable<string> GetSupportedExtensions(object model)
        {
            return (model as ISavableAs)?.GetSupportedExtensions();
        }

        public async Task Save(object model, IIOProvider provider)
        {
            var savable = model as ISavable;
            if (savable != null)
            {
                await savable.Save(provider);
            }
        }

        public async Task Save(object model, string filename, IIOProvider provider)
        {
            var savable = model as ISavableAs;
            if (savable != null)
            {
                await savable.Save(filename, provider);
            }
        }

        public bool SupportsSave(object model)
        {
            if (model is ISavableAs && model is IOnDisk)
            {
                return !string.IsNullOrEmpty((model as IOnDisk).Filename);
            }
            else
            {
                return model is ISavable;
            }            
        }

        public bool SupportsSaveAs(object model)
        {
            return model is ISavableAs;
        }
    }
}
